// @soyaxell09/zenbot-scraper/src/search/giphy.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

const TENOR_KEY = 'AIzaSyAyimkuYQYF_FXVALexPuGQctUWRURdCYQ'

function mapItem(g) {
  return {
    id:      g.id,
    title:   g.title || g.content_description || '',
    url:     g.itemurl || '',
    gif:     g.media_formats?.gif?.url      || g.media_formats?.mediumgif?.url || '',
    preview: g.media_formats?.nanogif?.url  || g.media_formats?.tinygif?.url  || '',
    mp4:     g.media_formats?.mp4?.url      || g.media_formats?.loopedmp4?.url || '',
    width:   g.media_formats?.gif?.dims?.[0] || 0,
    height:  g.media_formats?.gif?.dims?.[1] || 0,
  }
}

const _gifCache = new Map()

export async function giphy(query, limit = 5, type = 'search') {
  const endpoint = type === 'trending'
    ? `https://tenor.googleapis.com/v2/featured?key=${TENOR_KEY}&limit=${limit}&media_filter=gif`
    : `https://tenor.googleapis.com/v2/search?q=${encodeURIComponent(query)}&key=${TENOR_KEY}&limit=${limit}&media_filter=gif,mp4`

  const res  = await axios.get(endpoint, { headers: HEADERS, timeout: 15000 })
  const data = res.data?.results || []
  if (!data.length) throw new Error('Sin resultados en Tenor')
  return data.map(mapItem)
}

export async function gifSearch(query) {
  const cacheKey = `gif:${query}`
  let cached = _gifCache.get(cacheKey)
  if (!cached) {
    cached = { items: [], next: null, done: false }
    _gifCache.set(cacheKey, cached)
    setTimeout(() => _gifCache.delete(cacheKey), 1000 * 60 * 20)
  }
  return cached
}

export async function gifNext(query) {
  const cached = await gifSearch(query)
  if (cached.done && cached.items.length === 0) throw new Error('Sin resultados')

  if (!cached.done) {
    const params = {
      q:            query,
      key:          TENOR_KEY,
      limit:        20,
      media_filter: 'gif',
    }
    if (cached.next) params.pos = cached.next
    params.media_filter = 'gif,mp4'

    const res  = await axios.get('https://tenor.googleapis.com/v2/search', { params, headers: HEADERS, timeout: 15000 })
    const data = res.data?.results || []
    cached.next = res.data?.next || null
    if (!cached.next || !data.length) cached.done = true
    cached.items.push(...data.map(mapItem))
  }

  if (!cached.items.length) throw new Error('Sin resultados')

  const item = cached.items.shift()
  return item
}

export async function giphyBuffer(query) {
  const item = await gifNext(query)
  const url  = item.mp4 || item.gif
  if (!url) throw new Error('Sin URL de descarga')
  const res  = await axios.get(url, { responseType: 'arraybuffer', headers: HEADERS, timeout: 20000 })
  return {
    buffer:   Buffer.from(res.data),
    mimetype: item.mp4 ? 'video/mp4' : 'image/gif',
    title:    item.title,
    url,
  }
}
