// @soyaxell09/zenbot-scraper/src/search/stickersearch.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'
import * as cheerio from 'cheerio'

export async function stickerSearch(text, limit = 10) {
  try {
    const response = await axios.get(
      `https://getstickerpack.com/stickers?query=${encodeURIComponent(text)}`
    )
    const $     = cheerio.load(response.data)
    const packs = $('.sticker-pack-block')

    if (packs.length === 0)
      return { status: false, text: 'ningún resultado encontrado' }

    const randomIndex  = Math.floor(Math.random() * packs.length)
    const selectedPack = $(packs[randomIndex])
    const title        = selectedPack.find('.title').text().trim()
    const creator      = selectedPack.find('.username').text().trim()
    const packLink     = selectedPack.closest('a').attr('href')

    const packResponse = await axios.get(packLink)
    const packHtml     = cheerio.load(packResponse.data)
    const images       = packHtml('.sticker-pack-cols img')
    const totalImages  = images.length

    const imageArray = []
    images.each((i, img) => {
      if (i < limit) imageArray.push(packHtml(img).attr('src'))
    })

    return {
      status:  true,
      nombre:  title,
      creador: creator,
      total:   totalImages,
      fotos:   imageArray,
      url:     packLink,
    }
  } catch (error) {
    return { status: false, text: 'Error al acceder al sitio' }
  }
}
