// @soyaxell09/zenbot-scraper/src/tools/attp.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import gifencPkg from 'gifenc'
const { GIFEncoder, quantize, applyPalette } = gifencPkg

const FONT_7x9 = {
  'A': ['0011100','0100010','1000001','1000001','1111111','1000001','1000001','1000001','0000000'],
  'B': ['1111110','1000001','1000001','1111110','1000001','1000001','1000001','1111110','0000000'],
  'C': ['0111110','1000001','1000000','1000000','1000000','1000000','1000001','0111110','0000000'],
  'D': ['1111100','1000010','1000001','1000001','1000001','1000001','1000010','1111100','0000000'],
  'E': ['1111111','1000000','1000000','1111110','1000000','1000000','1000000','1111111','0000000'],
  'F': ['1111111','1000000','1000000','1111110','1000000','1000000','1000000','1000000','0000000'],
  'G': ['0111110','1000001','1000000','1000000','1001111','1000001','1000001','0111111','0000000'],
  'H': ['1000001','1000001','1000001','1111111','1000001','1000001','1000001','1000001','0000000'],
  'I': ['0111110','0001000','0001000','0001000','0001000','0001000','0001000','0111110','0000000'],
  'J': ['0001111','0000010','0000010','0000010','0000010','1000010','1000010','0111100','0000000'],
  'K': ['1000001','1000010','1000100','1001000','1110000','1001000','1000100','1000010','0000000'],
  'L': ['1000000','1000000','1000000','1000000','1000000','1000000','1000000','1111111','0000000'],
  'M': ['1000001','1100011','1010101','1001001','1000001','1000001','1000001','1000001','0000000'],
  'N': ['1000001','1100001','1010001','1001001','1000101','1000011','1000001','1000001','0000000'],
  'O': ['0111110','1000001','1000001','1000001','1000001','1000001','1000001','0111110','0000000'],
  'P': ['1111110','1000001','1000001','1111110','1000000','1000000','1000000','1000000','0000000'],
  'Q': ['0111110','1000001','1000001','1000001','1000001','1000101','1000010','0111101','0000000'],
  'R': ['1111110','1000001','1000001','1111110','1001000','1000100','1000010','1000001','0000000'],
  'S': ['0111110','1000001','1000000','0111110','0000001','0000001','1000001','0111110','0000000'],
  'T': ['1111111','0001000','0001000','0001000','0001000','0001000','0001000','0001000','0000000'],
  'U': ['1000001','1000001','1000001','1000001','1000001','1000001','1000001','0111110','0000000'],
  'V': ['1000001','1000001','1000001','1000001','1000001','0100010','0010100','0001000','0000000'],
  'W': ['1000001','1000001','1000001','1001001','1001001','1010101','1100011','1000001','0000000'],
  'X': ['1000001','0100010','0010100','0001000','0001000','0010100','0100010','1000001','0000000'],
  'Y': ['1000001','0100010','0010100','0001000','0001000','0001000','0001000','0001000','0000000'],
  'Z': ['1111111','0000001','0000010','0000100','0001000','0010000','0100000','1111111','0000000'],
  'a': ['0000000','0111110','0000001','0111111','1000001','1000011','0111101','0000000','0000000'],
  'b': ['1000000','1000000','1000000','1111110','1000001','1000001','1000001','1111110','0000000'],
  'c': ['0000000','0111110','1000001','1000000','1000000','1000001','1000001','0111110','0000000'],
  'd': ['0000001','0000001','0000001','0111111','1000001','1000001','1000001','0111111','0000000'],
  'e': ['0000000','0111110','1000001','1111111','1000000','1000000','1000001','0111110','0000000'],
  'f': ['0001110','0010001','0010000','0111100','0010000','0010000','0010000','0010000','0000000'],
  'g': ['0000000','0111111','1000001','1000001','0111111','0000001','0000001','0111110','0000000'],
  'h': ['1000000','1000000','1000000','1111110','1000001','1000001','1000001','1000001','0000000'],
  'i': ['0001000','0000000','0011000','0001000','0001000','0001000','0001000','0111110','0000000'],
  'j': ['0000100','0000000','0001100','0000100','0000100','0000100','1000100','0111000','0000000'],
  'k': ['1000000','1000000','1000010','1000100','1001000','1110000','1001000','1000110','0000000'],
  'l': ['0011000','0001000','0001000','0001000','0001000','0001000','0001000','0111110','0000000'],
  'm': ['0000000','0000000','1101101','1010101','1010101','1000001','1000001','1000001','0000000'],
  'n': ['0000000','0000000','1111110','1000001','1000001','1000001','1000001','1000001','0000000'],
  'o': ['0000000','0111110','1000001','1000001','1000001','1000001','1000001','0111110','0000000'],
  'p': ['0000000','1111110','1000001','1000001','1111110','1000000','1000000','1000000','0000000'],
  'q': ['0000000','0111111','1000001','1000001','0111111','0000001','0000001','0000001','0000000'],
  'r': ['0000000','0000000','1011110','1100001','1000000','1000000','1000000','1000000','0000000'],
  's': ['0000000','0111110','1000000','0111110','0000001','0000001','1000001','0111110','0000000'],
  't': ['0010000','0010000','1111110','0010000','0010000','0010000','0010001','0001110','0000000'],
  'u': ['0000000','0000000','1000001','1000001','1000001','1000001','1000011','0111101','0000000'],
  'v': ['0000000','0000000','1000001','1000001','0100010','0100010','0010100','0001000','0000000'],
  'w': ['0000000','0000000','1000001','1000001','1001001','1010101','1100011','1000001','0000000'],
  'x': ['0000000','0000000','1000001','0100010','0011100','0011100','0100010','1000001','0000000'],
  'y': ['0000000','1000001','1000001','0111111','0000001','0000001','1000001','0111110','0000000'],
  'z': ['0000000','0000000','1111111','0000010','0000100','0001000','0010000','1111111','0000000'],
  '0': ['0111110','1000001','1000011','1000101','1001001','1010001','1100001','0111110','0000000'],
  '1': ['0001000','0011000','0001000','0001000','0001000','0001000','0001000','0111110','0000000'],
  '2': ['0111110','1000001','0000001','0000110','0011000','0100000','1000000','1111111','0000000'],
  '3': ['0111110','1000001','0000001','0011110','0000001','0000001','1000001','0111110','0000000'],
  '4': ['0000100','0001100','0010100','0100100','1000100','1111111','0000100','0000100','0000000'],
  '5': ['1111111','1000000','1000000','1111110','0000001','0000001','1000001','0111110','0000000'],
  '6': ['0111110','1000001','1000000','1111110','1000001','1000001','1000001','0111110','0000000'],
  '7': ['1111111','0000001','0000010','0000100','0001000','0010000','0010000','0010000','0000000'],
  '8': ['0111110','1000001','1000001','0111110','1000001','1000001','1000001','0111110','0000000'],
  '9': ['0111110','1000001','1000001','0111111','0000001','0000001','1000001','0111110','0000000'],
  ' ': ['0000000','0000000','0000000','0000000','0000000','0000000','0000000','0000000','0000000'],
  '!': ['0001000','0001000','0001000','0001000','0001000','0000000','0000000','0001000','0000000'],
  '?': ['0111110','1000001','0000001','0000110','0001000','0001000','0000000','0001000','0000000'],
  '.': ['0000000','0000000','0000000','0000000','0000000','0000000','0000000','0001000','0000000'],
  ',': ['0000000','0000000','0000000','0000000','0000000','0001100','0001000','0010000','0000000'],
  '-': ['0000000','0000000','0000000','1111111','0000000','0000000','0000000','0000000','0000000'],
  ':': ['0000000','0001000','0000000','0000000','0000000','0001000','0000000','0000000','0000000'],
}

const CHAR_W   = 7
const CHAR_H   = 9
const CHAR_GAP = 1

const COLORS = [
  [255, 0,   0  ],
  [255, 127, 0  ],
  [255, 220, 0  ],
  [0,   210, 0  ],
  [0,   120, 255],
  [148, 0,   211],
]

function markPixel(mask, px, py, W, H) {
  if (px < 0 || py < 0 || px >= W || py >= H) return
  mask[py * W + px] = true
}

function setPixel(pixels, px, py, r, g, b, a, W, H) {
  if (px < 0 || py < 0 || px >= W || py >= H) return
  const i = (py * W + px) * 4
  pixels[i]     = r
  pixels[i + 1] = g
  pixels[i + 2] = b
  pixels[i + 3] = a
}

export async function attp(text, scale = 5) {
  if (!text || typeof text !== 'string') throw new Error('El texto es requerido')
  const str  = text.toUpperCase().slice(0, 40)
  const PAD  = 12 + scale
  const W    = str.length * (CHAR_W + CHAR_GAP) * scale + PAD * 2
  const H    = CHAR_H * scale + PAD * 2
  const gif  = GIFEncoder()

  for (const [fr, fg, fb] of COLORS) {
    const pixels     = new Uint8Array(W * H * 4)
    const letterMask = new Uint8Array(W * H)

    let cx = PAD
    for (const ch of str) {
      const glyph = FONT_7x9[ch] || FONT_7x9[' ']
      for (let row = 0; row < glyph.length; row++) {
        for (let col = 0; col < glyph[row].length; col++) {
          if (glyph[row][col] === '1') {
            for (let sy = 0; sy < scale; sy++)
              for (let sx = 0; sx < scale; sx++)
                markPixel(letterMask, cx + col * scale + sx, PAD + row * scale + sy, W, H)
          }
        }
      }
      cx += (CHAR_W + CHAR_GAP) * scale
    }

    const outlineThickness = Math.max(1, Math.round(scale * 0.35))
    for (let py = 0; py < H; py++) {
      for (let px = 0; px < W; px++) {
        if (letterMask[py * W + px]) continue
        let near = false
        outer: for (let dy = -outlineThickness; dy <= outlineThickness; dy++) {
          for (let dx = -outlineThickness; dx <= outlineThickness; dx++) {
            const nx = px + dx, ny = py + dy
            if (nx < 0 || ny < 0 || nx >= W || ny >= H) continue
            if (letterMask[ny * W + nx]) { near = true; break outer }
          }
        }
        if (near) setPixel(pixels, px, py, 0, 0, 0, 255, W, H)
      }
    }

    for (let py = 0; py < H; py++)
      for (let px = 0; px < W; px++)
        if (letterMask[py * W + px])
          setPixel(pixels, px, py, fr, fg, fb, 255, W, H)

    const palette = quantize(pixels, 256, { format: 'rgba4444' })
    const index   = applyPalette(pixels, palette, 'rgba4444')

    let transparentIndex = 0
    for (let i = 0; i < palette.length; i++) {
      if (palette[i][3] === 0) { transparentIndex = i; break }
    }

    gif.writeFrame(index, W, H, {
      palette,
      delay: 130,
      transparent: true,
      transparentIndex,
    })
  }

  gif.finish()
  return Buffer.from(gif.bytesView())
}
