// @soyaxell09/zenbot-scraper/src/tools/lyrics.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

async function tryLyricsOvh(artist, title) {
  const res = await axios.get(
    `https://api.lyrics.ovh/v1/${encodeURIComponent(artist)}/${encodeURIComponent(title)}`,
    { headers: HEADERS, timeout: 10000 }
  )
  if (!res.data?.lyrics) throw new Error('Sin letra')
  return res.data.lyrics.trim()
}

async function tryLrclib(query) {
  const res = await axios.get(
    `https://lrclib.net/api/search?q=${encodeURIComponent(query)}`,
    { headers: HEADERS, timeout: 10000 }
  )
  const results = res.data
  if (!results?.length) throw new Error('Sin resultados')
  return results.map(r => ({
    id:        r.id,
    title:     r.trackName || r.name || '',
    artist:    r.artistName || '',
    album:     r.albumName  || '',
    duration:  r.duration   || 0,
    lyrics:    r.plainLyrics || null,
    lrc:       r.syncedLyrics || null,
  }))
}

export async function lyricsSearch(query, limit = 5) {
  const errors = []

  try {
    const results = await tryLrclib(query)
    return results.slice(0, limit)
  } catch (e) { errors.push('lrclib: ' + e.message) }

  const parts  = query.trim().split(/\s+/)
  const mid    = Math.ceil(parts.length / 2)
  const artist = parts.slice(0, mid).join(' ')
  const title  = parts.slice(mid).join(' ') || artist

  try {
    const lyrics = await tryLyricsOvh(artist, title)
    return [{
      id:       null,
      title:    title,
      artist:   artist,
      album:    '',
      duration: 0,
      lyrics,
      lrc:      null,
    }]
  } catch (e) { errors.push('lyrics.ovh: ' + e.message) }

  throw new Error('No se encontraron letras. Errores: ' + errors.join(' | '))
}

export async function lyricsGet(artist, title) {
  const errors = []

  try {
    const results = await tryLrclib(`${artist} ${title}`)
    const match   = results.find(r =>
      r.title.toLowerCase().includes(title.toLowerCase()) ||
      r.artist.toLowerCase().includes(artist.toLowerCase())
    ) || results[0]
    if (match) return match
  } catch (e) { errors.push('lrclib: ' + e.message) }

  try {
    const lyrics = await tryLyricsOvh(artist, title)
    return { id: null, title, artist, album: '', duration: 0, lyrics, lrc: null }
  } catch (e) { errors.push('lyrics.ovh: ' + e.message) }

  throw new Error('No se encontró la letra. Errores: ' + errors.join(' | '))
}
