// @soyaxell09/zenbot-scraper/src/tools/news.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

const FEEDS = {
  es:      'https://feeds.bbci.co.uk/mundo/rss.xml',
  en:      'https://feeds.bbci.co.uk/news/rss.xml',
  pt:      'https://feeds.bbci.co.uk/portuguese/rss.xml',
  tech:    'https://feeds.feedburner.com/TechCrunch',
  sports:  'https://www.espn.com/espn/rss/news',
  science: 'https://feeds.feedburner.com/sciencedaily',
  world:   'https://feeds.bbci.co.uk/news/world/rss.xml',
}

function parseRss2Json(data, limit) {
  return (data.items || []).slice(0, limit).map(item => ({
    title:       item.title || '',
    description: item.description?.replace(/<[^>]+>/g, '').slice(0, 200) || '',
    url:         item.link  || '',
    image:       item.thumbnail || item.enclosure?.link || '',
    published:   item.pubDate || '',
    source:      data.feed?.title || '',
  }))
}

export async function news(category = 'es', limit = 5) {
  const feedUrl = FEEDS[category] || FEEDS.es

  const res = await axios.get(
    `https://rss2json.com/api.json?rss_url=${encodeURIComponent(feedUrl)}&count=${limit}`,
    { headers: HEADERS, timeout: 15000 }
  )

  if (res.data?.status !== 'ok') throw new Error('Error al obtener noticias')

  return {
    category,
    source:  res.data.feed?.title || '',
    items:   parseRss2Json(res.data, limit),
  }
}

export function newsCategories() {
  return Object.keys(FEEDS)
}
