// @soyaxell09/zenbot-scraper/src/tools/tiktokstalk.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

function formatNumber(n) {
  if (!n) return '0'
  if (n >= 1_000_000_000) return (n / 1_000_000_000).toFixed(1).replace(/\.0$/, '') + 'B'
  if (n >= 1_000_000)     return (n / 1_000_000).toFixed(1).replace(/\.0$/, '') + 'M'
  if (n >= 1_000)         return (n / 1_000).toFixed(1).replace(/\.0$/, '') + 'K'
  return String(n)
}

export async function tiktokStalk(username) {
  const clean = username.replace(/^@/, '').trim()
  if (!clean) throw new Error('Username inválido')

  const res = await axios.get(
    `https://tikwm.com/api/user/info?unique_id=${encodeURIComponent(clean)}`,
    { headers: HEADERS, timeout: 15000 }
  )

  const data = res.data
  if (data?.code !== 0 || !data?.data) throw new Error('tikwm: ' + (data?.msg || 'usuario no encontrado'))

  const u = data.data.user
  const s = data.data.stats

  const followers  = s.followerCount  || 0
  const following  = s.followingCount || 0
  const likes      = s.heartCount     || 0
  const videos     = s.videoCount     || 0
  const avgLikes   = videos > 0 ? Math.round(likes / videos) : 0

  return {
    id:          u.id       || '',
    username:    u.uniqueId || '',
    nickname:    u.nickname || '',
    bio:         u.signature || '',
    avatar:      u.avatarLarger || u.avatarMedium || u.avatarThumb || '',
    verified:    u.verified || false,
    private:     u.privateAccount || false,
    bioLink:     u.bioLink?.link || null,
    instagram:   u.ins_id || null,
    twitter:     u.twitter_id || null,
    youtube:     u.youtube_channel_title || null,
    createdAt:   u.createTime ? new Date(u.createTime * 1000).toISOString().slice(0, 10) : null,
    stats: {
      followers:      followers,
      following:      following,
      likes:          likes,
      videos:         videos,
      followersStr:   formatNumber(followers),
      followingStr:   formatNumber(following),
      likesStr:       formatNumber(likes),
      avgLikesStr:    formatNumber(avgLikes),
    },
    url: `https://www.tiktok.com/@${u.uniqueId}`,
  }
}
