// @soyaxell09/zenbot-scraper/src/tools/translator.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

const LANGS = {
  es: 'Español', en: 'English', pt: 'Português', fr: 'Français',
  de: 'Deutsch', it: 'Italiano', ja: '日本語', ko: '한국어',
  zh: '中文', ru: 'Русский', ar: 'العربية', hi: 'हिन्दी',
  tr: 'Türkçe', nl: 'Nederlands', pl: 'Polski', sv: 'Svenska',
  uk: 'Українська', vi: 'Tiếng Việt', id: 'Bahasa Indonesia',
}

async function tryGoogle(text, from, to) {
  const res = await axios.get(
    `https://translate.googleapis.com/translate_a/single?client=gtx&sl=${from}&tl=${to}&dt=t&q=${encodeURIComponent(text)}`,
    { headers: HEADERS, timeout: 10000 }
  )
  const data = res.data
  if (!Array.isArray(data?.[0])) throw new Error('Respuesta inválida')
  const translated = data[0].map(s => s[0]).filter(Boolean).join('')
  if (!translated) throw new Error('Sin traducción')
  const detectedLang = data[2] || from
  return { translated, detectedLang }
}

async function tryMyMemory(text, from, to) {
  const res = await axios.get(
    `https://api.mymemory.translated.net/get?q=${encodeURIComponent(text)}&langpair=${from}|${to}`,
    { headers: HEADERS, timeout: 10000 }
  )
  const translated = res.data?.responseData?.translatedText
  if (!translated) throw new Error('Sin traducción')
  return { translated, detectedLang: from }
}

export async function translate(text, to = 'es', from = 'auto') {
  if (!text?.trim()) throw new Error('Texto vacío')

  const errors = []

  try {
    const result = await tryGoogle(text, from, to)
    return {
      original:     text,
      translated:   result.translated,
      from:         result.detectedLang,
      to,
      fromName:     LANGS[result.detectedLang] || result.detectedLang,
      toName:       LANGS[to] || to,
      source:       'google',
    }
  } catch (e) { errors.push('google: ' + e.message) }

  try {
    const langFrom = from === 'auto' ? 'en' : from
    const result   = await tryMyMemory(text, langFrom, to)
    return {
      original:   text,
      translated: result.translated,
      from:       langFrom,
      to,
      fromName:   LANGS[langFrom] || langFrom,
      toName:     LANGS[to] || to,
      source:     'mymemory',
    }
  } catch (e) { errors.push('mymemory: ' + e.message) }

  throw new Error('No se pudo traducir. Errores: ' + errors.join(' | '))
}

export function getLangs() {
  return Object.entries(LANGS).map(([code, name]) => ({ code, name }))
}
