// @soyaxell09/zenbot-scraper/src/tools/urlshortener.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

async function tryIsGd(url) {
  const res = await axios.get(
    `https://is.gd/create.php?format=json&url=${encodeURIComponent(url)}`,
    { headers: HEADERS, timeout: 10000 }
  )
  if (!res.data?.shorturl) throw new Error('Sin resultado')
  return { short: res.data.shorturl, source: 'is.gd' }
}

async function tryTinyUrl(url) {
  const res = await axios.get(
    `https://tinyurl.com/api-create.php?url=${encodeURIComponent(url)}`,
    { headers: HEADERS, timeout: 10000 }
  )
  if (!res.data?.startsWith('http')) throw new Error('Sin resultado')
  return { short: res.data.trim(), source: 'tinyurl' }
}

export async function shortenUrl(url) {
  if (!url?.trim()) throw new Error('URL vacía')
  if (!url.startsWith('http')) throw new Error('URL inválida — debe empezar con http/https')

  const errors = []
  try { return await tryIsGd(url) }    catch (e) { errors.push('is.gd: '    + e.message) }
  try { return await tryTinyUrl(url) } catch (e) { errors.push('tinyurl: '  + e.message) }

  throw new Error('No se pudo acortar. Errores: ' + errors.join(' | '))
}

export async function expandUrl(shortUrl) {
  if (!shortUrl?.trim()) throw new Error('URL vacía')
  const res = await axios.get(shortUrl, {
    headers: HEADERS,
    maxRedirects: 10,
    timeout: 10000,
  })
  return {
    original: shortUrl,
    expanded: res.request?.res?.responseUrl || res.config?.url || shortUrl,
  }
}
