// @soyaxell09/zenbot-scraper/src/tools/weather.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

const WEATHER_CODES = {
  113: '☀️ Despejado', 116: '⛅ Parcialmente nublado', 119: '☁️ Nublado',
  122: '☁️ Cubierto', 143: '🌫️ Neblina', 176: '🌦️ Lluvia dispersa',
  179: '🌨️ Nevada dispersa', 182: '🌧️ Aguanieve', 185: '🌧️ Llovizna helada',
  200: '⛈️ Tormenta', 227: '🌨️ Ventisca', 230: '❄️ Ventisca fuerte',
  248: '🌫️ Niebla', 260: '🌫️ Niebla helada', 263: '🌦️ Llovizna',
  266: '🌧️ Llovizna ligera', 281: '🌧️ Llovizna helada', 284: '🌧️ Llovizna helada fuerte',
  293: '🌦️ Lluvia ligera', 296: '🌧️ Lluvia ligera', 299: '🌧️ Lluvia moderada',
  302: '🌧️ Lluvia moderada', 305: '🌧️ Lluvia fuerte', 308: '🌧️ Lluvia muy fuerte',
  311: '🌧️ Aguanieve ligera', 314: '🌧️ Aguanieve moderada', 317: '🌧️ Aguanieve',
  320: '🌨️ Nevada ligera', 323: '🌨️ Nevada ligera', 326: '🌨️ Nevada moderada',
  329: '❄️ Nevada moderada', 332: '❄️ Nevada fuerte', 335: '❄️ Nevada fuerte',
  338: '❄️ Nevada muy fuerte', 350: '🌧️ Granizo', 353: '🌦️ Lluvia ligera',
  356: '🌧️ Lluvia fuerte', 359: '🌧️ Lluvia torrencial', 362: '🌧️ Aguanieve ligera',
  365: '🌧️ Aguanieve moderada', 368: '🌨️ Nevada ligera', 371: '❄️ Nevada moderada',
  374: '🌧️ Granizo ligero', 377: '🌧️ Granizo', 386: '⛈️ Tormenta ligera',
  389: '⛈️ Tormenta fuerte', 392: '⛈️ Tormenta con nevada', 395: '❄️ Tormenta de nieve',
}

export async function weather(location) {
  if (!location?.trim()) throw new Error('Ubicación vacía')

  const res  = await axios.get(
    `https://wttr.in/${encodeURIComponent(location)}?format=j1`,
    { headers: HEADERS, timeout: 15000 }
  )

  const data    = res.data?.data || res.data
  const current = data.weather?.[0]?.hourly?.[0] || {}
  const now     = data.current_condition?.[0] || {}
  const req     = data.nearest_area?.[0]

  const city    = req?.areaName?.[0]?.value || location
  const country = req?.country?.[0]?.value || ''
  const code    = parseInt(now.weatherCode || 113)
  const desc    = WEATHER_CODES[code] || now.weatherDesc?.[0]?.value || ''

  return {
    location:    `${city}${country ? ', ' + country : ''}`,
    temp:        parseInt(now.temp_C || 0),
    feelsLike:   parseInt(now.FeelsLikeC || 0),
    humidity:    parseInt(now.humidity || 0),
    wind:        parseInt(now.windspeedKmph || 0),
    windDir:     now.winddir16Point || '',
    visibility:  parseInt(now.visibility || 0),
    pressure:    parseInt(now.pressure || 0),
    uvIndex:     parseInt(now.uvIndex || 0),
    description: desc,
    code,
    forecast:    (data.weather || []).map(d => ({
      date:    d.date,
      maxTemp: parseInt(d.maxtempC),
      minTemp: parseInt(d.mintempC),
      desc:    WEATHER_CODES[parseInt(d.hourly?.[4]?.weatherCode)] || d.hourly?.[4]?.weatherDesc?.[0]?.value || '',
      sunrise: d.astronomy?.[0]?.sunrise || '',
      sunset:  d.astronomy?.[0]?.sunset  || '',
    })),
  }
}
