// @soyaxell09/zenbot-scraper/src/scrapers/aptoide.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const API_BASE = 'https://ws75.aptoide.com/api/7'
const UA = 'Mozilla/5.0 (Linux; Android 11; Redmi Note 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36'

function extractSize(app) {
  const bytes = app.file?.filesize || app.file?.size || app.size || null
  if (!bytes || bytes <= 0) return 'N/A'
  if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(1) + ' KB'
  return (bytes / (1024 * 1024)).toFixed(2) + ' MB'
}

export async function aptoideSearch(query, limit = 10) {
  try {
    const { data } = await axios.get(`${API_BASE}/apps/search`, {
      params: { query, limit },
      headers: { 'User-Agent': UA },
      timeout: 10000
    })

    const list = data.datalist?.list || []
    return list.slice(0, limit).map(app => ({
      title: app.name,
      id: app.package,
      url: `https://es.aptoide.com/app/${app.package}`,
      thumb: app.icon,
      version: app.file?.vername || 'N/A',
      size: extractSize(app),
      rating: app.stats?.rating?.avg?.toFixed(1) || '0.0'
    }))
  } catch {
    return []
  }
}

export async function aptoideInfo(packageName) {
  try {
    const { data } = await axios.get(`${API_BASE}/app/get`, {
      params: { package_name: packageName },
      headers: { 'User-Agent': UA },
      timeout: 10000
    })

    const app = data.nodes?.meta?.data || data.datalist?.list?.[0]
    if (!app) return null

    const rawDesc = app.media?.description || app.description || ''

    return {
      title: app.name,
      id: app.package,
      version: app.file?.vername || 'N/A',
      size: extractSize(app),
      thumb: app.icon,
      updated: app.updated,
      downloads: app.stats?.downloads?.toLocaleString() || '0',
      rating: app.stats?.rating?.avg?.toFixed(1) || '0.0',
      total_reviews: app.stats?.rating?.total || 0,
      is_safe: app.file?.malware?.rank === 'TRUSTED',
      security_reason: app.file?.malware?.reason?.sig || 'Verificado',
      min_android: app.file?.hardware?.sdk ? `Android ${app.file.hardware.sdk}+` : 'Varía',
      arch: app.file?.hardware?.cpus?.join(', ') || 'Universal',
      md5: app.file?.md5sum,
      screenshots: app.media?.screenshots?.map(s => s.url) || [],
      video: app.media?.videos?.[0]?.url || null,
      description: rawDesc.replace(/<[^>]*>/g, '').trim().slice(0, 1000),
      download: app.file?.path || null,
      url: `https://es.aptoide.com/app/${app.package}`
    }
  } catch {
    return null
  }
}
