// @soyaxell09/zenbot-scraper/src/scrapers/books.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'
import * as cheerio from 'cheerio'

const DOMAINS = ['https://ww3.lectulandia.co', 'https://www.lectulandia.co']
const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'
const HDRS = {
  'User-Agent': UA,
  'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
  'Accept-Language': 'es-ES,es;q=0.9,en;q=0.8',
}

async function resolveAntUpload(url) {
  if (!url || !url.includes('antupload.com/file/')) return null
  try {
    const res = await axios.get(url, { headers: HDRS, timeout: 15000 })
    const cookies = res.headers['set-cookie']?.map(c => c.split(';')[0]).join('; ') || ''
    const $ = cheerio.load(res.data)
    
    const finalHref = $('#downloadB').attr('href') || $('a[href*="/filed/"]').first().attr('href') || $('.download-btn').attr('href')
    
    if (finalHref) {
      const absoluteLink = finalHref.startsWith('http') ? finalHref : 'https://www.antupload.com' + finalHref
      return {
        url: absoluteLink,
        headers: {
          ...HDRS,
          'Referer': url,
          'Cookie': cookies
        }
      }
    }
  } catch (error) {
    console.log('[SCRAPER] AntUpload Error:', error.message)
  }
  return null
}

async function fetchLectulandia(endpoint) {
  let lastError = null
  for (const base of DOMAINS) {
    const targetUrl = endpoint.startsWith('http') ? endpoint : `${base}${endpoint}`
    try {
      const res = await axios.get(targetUrl, { headers: HDRS, timeout: 12000 })
      return { data: res.data, base }
    } catch (e) {
      lastError = e
      console.log(`[SCRAPER] Bloqueo en ${base}: ${e.response?.status || e.message}. Usando proxies...`)
      
      const proxies = [
        `https://api.codetabs.com/v1/proxy?quest=${encodeURIComponent(targetUrl)}`,
        `https://api.allorigins.win/get?url=${encodeURIComponent(targetUrl)}`
      ]
      
      for (const proxy of proxies) {
        try {
          const resProxy = await axios.get(proxy, { timeout: 15000 })
          let html = resProxy.data
          if (proxy.includes('allorigins') && html.contents) html = html.contents
          
          if (html && (html.includes('lectulandia') || html.includes('article'))) {
            console.log(`[SCRAPER] Proxy exitoso!`)
            return { data: html, base }
          }
        } catch (err) {}
      }
    }
  }
  throw new Error(`Cloudflare bloqueó todas las rutas y proxies. Error original: ${lastError?.response?.status || lastError?.message}`)
}

export async function bookSearch(query, limit = 10) {
  const { data, base } = await fetchLectulandia(`/search/${encodeURIComponent(query)}`)
  const $ = cheerio.load(data)
  const results = []

  $('article.card').each((i, el) => {
    if (i >= limit) return false
    
    let href = $(el).find('a.card-click-target').attr('href') || $(el).find('a').first().attr('href') || ''
    if (href && !href.startsWith('http')) href = href.startsWith('/') ? href : '/' + href
    
    const url   = href.startsWith('http') ? href : base + href
    const title = $(el).find('img.cover').attr('alt')?.trim() || $(el).find('h2 a').text().trim() || $(el).find('h2').text().trim()
    const thumb = $(el).find('img.cover').attr('src') || ''
    const author = $(el).find('a[href*="/autor/"]').first().text().trim()
    const genre  = $(el).find('a[href*="/genero/"]').first().text().trim() || null

    if (!title || !url || url === base) return
    results.push({ title, author, thumb, url, genre })
  })

  if (!results.length) throw new Error('Lectulandia no devolvió resultados o Cloudflare bloqueó la IP.')
  return results
}

export async function bookInfo(url) {
  const { data, base } = await fetchLectulandia(url)
  const $ = cheerio.load(data)

  const title  = $('h1[itemprop="name"]').text().trim() || $('#title').text().trim() || $('h1').text().trim()
  const author = $('#autor').text().replace(/^Autor:\s*/i, '').trim() || $('a[href*="/autor/"]').first().text().trim()
  const genre  = $('a[href*="/genero/"]').first().text().replace(/^Genero:\s*/i, '').trim() || null
  const year   = $('#publicado').text().replace(/^Publicado:\s*/i, '').trim() || null

  let description = ''
  $('#book .description, #content .description, .book .description, #resumen').each((_, el) => {
    const txt = $(el).text().trim()
    if (txt.length > description.length) description = txt
  })
  if (!description) description = $('meta[property="og:description"]').attr('content')?.trim() || ''

  let thumb = $('meta[property="og:image"]').attr('content') || $('img.cover').attr('src') || ''
  if (thumb && !thumb.startsWith('http')) thumb = thumb.startsWith('/') ? base + thumb : base + '/' + thumb

  const getAntUploadLink = (href) => {
    if (!href) return null
    const dMatch = href.match(/[?&]d=([^&]+)/)
    if (dMatch && dMatch[1]) {
      try {
        const decoded = Buffer.from(dMatch[1], 'base64').toString('utf-8')
        return 'https://www.antupload.com/file/' + decoded
      } catch (e) {}
    }
    return href.startsWith('http') ? href : base + href
  }

  let epubLink = null
  let pdfLink  = null

  $('a').each((_, el) => {
    const href = $(el).attr('href') || ''
    if (href.includes('download.php') || href.includes('?d=')) {
      const tVal = href.match(/[?&]t=(\d+)/)?.[1]
      if (tVal === '1') epubLink = getAntUploadLink(href)
      if (tVal === '2') pdfLink  = getAntUploadLink(href)
    }
  })

  if (!epubLink) epubLink = getAntUploadLink($('a#download1').attr('href') || $('a:contains("EPUB")').attr('href') || '')
  if (!pdfLink)  pdfLink  = getAntUploadLink($('a#download2').attr('href') || $('a:contains("PDF")').attr('href') || '')

  const [finalEpub, finalPdf] = await Promise.all([
    resolveAntUpload(epubLink),
    resolveAntUpload(pdfLink)
  ])

  return {
    title,
    author,
    description: description.replace(/\s+/g, ' ').slice(0, 1000),
    genre,
    year,
    thumb,
    url,
    download: {
      epub: finalEpub || null,
      pdf:  finalPdf  || null,
    }
  }
}
