// @soyaxell09/zenbot-scraper/src/scrapers/ddownr.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'
const API_KEY = 'dfcb6d76f2f6a9894gjkege8a4ab23'

async function safeGet(url, config = {}) {
  try {
    return await axios.get(url, config)
  } catch (e) {
    const proxies = [
      `https://api.codetabs.com/v1/proxy?quest=${encodeURIComponent(url)}`,
      `https://api.allorigins.win/raw?url=${encodeURIComponent(url)}`
    ]
    for (const p of proxies) {
      try { return await axios.get(p, config) } catch (err) {}
    }
    throw e
  }
}

export const ddownrFormats = {
  video: ['360', '480', '720', '1080', '1440', '4k', '8k'],
  audio: ['mp3', 'm4a', 'webm', 'aac', 'flac', 'opus', 'ogg', 'wav']
}

export async function ddownr(url, format = '720') {
  try {
    const requestedFormat = format.toString().toLowerCase()
    const isValidFormat = ddownrFormats.video.includes(requestedFormat) || ddownrFormats.audio.includes(requestedFormat)

    if (!isValidFormat) {
      throw new Error(`Formato no soportado. Video: ${ddownrFormats.video.join(', ')} | Audio: ${ddownrFormats.audio.join(', ')}`)
    }

    const initUrl = `https://p.savenow.to/ajax/download.php?copyright=0&format=${requestedFormat}&url=${encodeURIComponent(url)}&api=${API_KEY}`
    
    const initRes = await safeGet(initUrl, { headers: { 'User-Agent': UA }, timeout: 15000 })

    if (!initRes.data || !initRes.data.success || !initRes.data.id) {
      throw new Error('No se pudo iniciar la tarea de descarga (Rechazo del servidor Savenow).')
    }

    const { id, title, info } = initRes.data
    const thumb = info?.image || ''
    
    let downloadUrl = null
    let attempts = 0
    const maxAttempts = 80 

    while (attempts < maxAttempts) {
      await new Promise(resolve => setTimeout(resolve, 3000))
      
      const progRes = await safeGet(`https://p.savenow.to/ajax/progress.php?id=${id}`, {
        headers: { 'User-Agent': UA },
        timeout: 10000
      })

      const { success, download_url, text } = progRes.data

      if (text === 'Error' || text === 'Failed') {
        throw new Error('El servidor de conversión Savenow falló al procesar este video (posible bloqueo por Copyright).')
      }

      if (success === 1 && download_url) {
        downloadUrl = download_url
        break
      }

      attempts++
    }

    if (!downloadUrl) {
      throw new Error('El servidor tardó demasiado en procesar el video (Timeout).')
    }

    return {
      title: title || info?.title || 'YouTube Video',
      thumb: thumb,
      format: requestedFormat,
      download: downloadUrl
    }

  } catch (error) {
    throw new Error(`Ddownr Error: ${error.message}`)
  }
}
