// @soyaxell09/zenbot-scraper/src/scrapers/deezer.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 */

import axios from 'axios'

const UA      = 'Mozilla/5.0 (Linux; Android 11; Redmi Note 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36'
const BASE    = 'https://flacdownloader.com'
const APIBASE = `${BASE}/flac`
const KEY     = 'l@p*gute)77=g5clebcp4lz#=x%(*rwg+ku0_)bh=&%6wg!a'

const HEADERS = {
  'User-Agent': UA,
  'Referer': `${BASE}/`,
}

export async function deezerSearch(query, limit = 10) {
  const { data } = await axios.get(`${APIBASE}/search`, {
    params: { query },
    headers: HEADERS,
    timeout: 12000,
  })

  return (data.data || []).slice(0, limit).map(t => ({
    id:       t.id,
    title:    t.title,
    duration: t.duration,
    preview:  t.preview,
    explicit: t.explicit_lyrics,
    link:     t.link,
    artist: {
      id:     t.artist?.id,
      name:   t.artist?.name,
      picture: t.artist?.picture_medium,
    },
    album: {
      id:    t.album?.id,
      title: t.album?.title,
      cover: t.album?.cover_medium,
      cover_xl: t.album?.cover_xl,
    },
  }))
}

export async function deezerDownload(trackId, format = 'flac') {
  const { data: tokenData } = await axios.get(`${APIBASE}/download-token`, {
    params: { t: trackId, f: format },
    headers: { ...HEADERS, 'X-Download-Access': KEY },
    timeout: 10000,
  })

  const { token, expires } = tokenData

  const dlRes = await axios.get(`${APIBASE}/download`, {
    params: {
      t: trackId,
      f: format,
      token,
      expires,
    },
    headers: HEADERS,
    responseType: 'arraybuffer',
    timeout: 60000,
  })

  const ext      = format.toLowerCase().includes('flac') ? 'flac' : 'mp3'
  const mimeType = ext === 'flac' ? 'audio/flac' : 'audio/mpeg'

  return {
    buffer:   Buffer.from(dlRes.data),
    ext,
    mimeType,
    size:     parseInt(dlRes.headers['content-length'] || '0', 10),
  }
}

export async function deezerTrack(query, format = 'mp3') {
  const results = await deezerSearch(query, 1)
  if (!results.length) throw new Error('No se encontraron resultados.')

  const track    = results[0]
  const download = await deezerDownload(track.id, format)

  return {
    title:    track.title,
    artist:   track.artist.name,
    album:    track.album.title,
    duration: track.duration,
    cover:    track.album.cover_xl,
    preview:  track.preview,
    link:     track.link,
    explicit: track.explicit,
    download,
  }
}
