// @soyaxell09/zenbot-scraper/src/scrapers/gdrive.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

function parseFileId(input) {
  const m1 = input.match(/\/d\/([a-zA-Z0-9_-]{20,})/)
  if (m1) return m1[1]
  const m2 = input.match(/[?&]id=([a-zA-Z0-9_-]{20,})/)
  if (m2) return m2[1]
  if (/^[a-zA-Z0-9_-]{20,}$/.test(input)) return input
  return null
}

export async function gdriveInfo(input) {
  const fileId = parseFileId(input)
  if (!fileId) throw new Error('URL o ID de Google Drive inválido')

  const res   = await axios.get(`https://drive.google.com/file/d/${fileId}/view`, { headers: HEADERS, timeout: 15000 })
  const html  = res.data
  const title = html.match(/<title>([^<]+)<\/title>/)?.[1]?.replace(' - Google Drive', '').trim() || ''

  return {
    fileId,
    name:     title,
    download: `https://drive.google.com/uc?export=download&id=${fileId}`,
    url:      `https://drive.google.com/file/d/${fileId}/view`,
  }
}

export async function gdriveDownload(input) {
  const fileId = parseFileId(input)
  if (!fileId) throw new Error('URL o ID de Google Drive inválido')

  const res = await axios.get(
    `https://drive.google.com/uc?export=download&id=${fileId}`,
    { headers: HEADERS, maxRedirects: 10, timeout: 30000, responseType: 'arraybuffer' }
  )

  const contentType = res.headers['content-type'] || 'application/octet-stream'
  const finalUrl    = res.request?.res?.responseUrl || `https://drive.usercontent.google.com/download?id=${fileId}&export=download`

  if (contentType.includes('text/html')) {
    const html    = Buffer.from(res.data).toString()
    const confirm = html.match(/confirm=([^&"]+)/)
    if (confirm) {
      const res2 = await axios.get(
        `https://drive.google.com/uc?export=download&id=${fileId}&confirm=${confirm[1]}`,
        { headers: HEADERS, maxRedirects: 10, timeout: 30000, responseType: 'arraybuffer' }
      )
      return {
        fileId,
        buffer:      Buffer.from(res2.data),
        contentType: res2.headers['content-type'] || contentType,
        size:        parseInt(res2.headers['content-length'] || '0'),
        url:         finalUrl,
      }
    }
    throw new Error('Google Drive requiere autenticación o el archivo no es público')
  }

  return {
    fileId,
    buffer:      Buffer.from(res.data),
    contentType,
    size:        parseInt(res.headers['content-length'] || '0'),
    url:         finalUrl,
  }
}
