// @soyaxell09/zenbot-scraper/src/scrapers/mediafire.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'
import * as cheerio from 'cheerio'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
}

function parseKey(url) {
  const m = url.match(/mediafire\.com\/file\/([a-z0-9]+)/)
  return m ? m[1] : null
}

function deduplicateName(raw) {
  const clean = raw.trim().replace(/\s+/g, ' ')
  const half  = Math.ceil(clean.length / 2)
  const first = clean.slice(0, half)
  const second = clean.slice(half).trim()
  return second.startsWith(first.trim()) ? first.trim() : clean
}

async function tryPage(url) {
  const res  = await axios.get(url, { headers: HEADERS, timeout: 15000 })
  const $    = cheerio.load(res.data)
  const link = $('a#downloadButton').attr('href') || $('a.input').attr('href')
  if (!link) throw new Error('page: sin link de descarga')
  const name = deduplicateName($('div.filename').text())
  const size = $('ul.details li').first().text().replace('File size:', '').trim()
  return { link, name, size }
}

async function tryAPI(key) {
  const res  = await axios.get(
    `https://www.mediafire.com/api/1.5/file/get_links.php?quick_key=${key}&link_type=normal_download&response_format=json`,
    { headers: HEADERS, timeout: 15000 }
  )
  const data = res.data?.response
  if (data?.result !== 'Success') throw new Error('api: ' + (data?.message || 'sin resultado'))
  const dl = data?.links?.[0]?.normal_download
  if (!dl) throw new Error('api: sin download link')
  return dl
}

export async function mediafireInfo(url) {
  if (!url.includes('mediafire.com')) throw new Error('URL de MediaFire inválida')

  const key    = parseKey(url)
  const errors = []
  let   info   = null

  try {
    info = await tryPage(url)
  } catch (e) { errors.push('page: ' + e.message) }

  if (!info?.link && key) {
    try {
      const link = await tryAPI(key)
      info = { ...(info || {}), link }
    } catch (e) { errors.push('api: ' + e.message) }
  }

  if (!info?.link) throw new Error('No se pudo obtener el link. Errores: ' + errors.join(' | '))

  return {
    key:      key || '',
    name:     info.name || '',
    size:     info.size || '',
    download: info.link,
    url,
  }
}
