// @soyaxell09/zenbot-scraper/src/scrapers/twitter.js

/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
  'Referer':    'https://platform.twitter.com/',
}

function parseTweetId(url) {
  const m = url.match(/\/status\/(\d+)/)
  return m ? m[1] : (/^\d+$/.test(url) ? url : null)
}

function parseMedia(details) {
  if (!details?.length) return []
  return details.map(m => {
    if (m.type === 'video' || m.type === 'animated_gif') {
      const variants = m.video_info?.variants || []
      const best     = variants
        .filter(v => v.content_type === 'video/mp4')
        .sort((a, b) => (b.bitrate || 0) - (a.bitrate || 0))[0]
      return {
        type:      m.type === 'animated_gif' ? 'gif' : 'video',
        url:       best?.url || '',
        thumbnail: m.media_url_https + '?format=jpg&name=large',
        width:     m.original_info?.width  || 0,
        height:    m.original_info?.height || 0,
        variants:  variants.filter(v => v.url).map(v => ({
          url:         v.url,
          contentType: v.content_type,
          bitrate:     v.bitrate || 0,
        })),
      }
    }
    return {
      type:   'photo',
      url:    m.media_url_https + '?format=jpg&name=orig',
      thumb:  m.media_url_https + '?format=jpg&name=small',
      width:  m.original_info?.width  || 0,
      height: m.original_info?.height || 0,
    }
  })
}

export async function tweetInfo(url) {
  const id = parseTweetId(url)
  if (!id) throw new Error('URL/ID de Twitter inválido')

  const res = await axios.get(
    `https://cdn.syndication.twimg.com/tweet-result?id=${id}&lang=en&features=tfw_timeline_list%3A%3Btfw_follower_count_sunset%3Atrue&token=abc123`,
    { headers: HEADERS, timeout: 15000 }
  )

  const d = res.data
  if (!d?.id_str) throw new Error('Tweet no encontrado o privado')

  const medias = parseMedia(d.mediaDetails)

  return {
    id:         d.id_str,
    text:       d.text || '',
    lang:       d.lang || '',
    createdAt:  d.created_at || '',
    likes:      d.favorite_count || 0,
    replies:    d.conversation_count || 0,
    author: {
      id:       d.user?.id_str || '',
      name:     d.user?.name || '',
      username: d.user?.screen_name || '',
      avatar:   d.user?.profile_image_url_https?.replace('_normal', '') || '',
      verified: d.user?.is_blue_verified || false,
    },
    hashtags:   d.entities?.hashtags?.map(h => h.text) || [],
    mentions:   d.entities?.user_mentions?.map(m => m.screen_name) || [],
    urls:       d.entities?.urls?.map(u => u.expanded_url) || [],
    medias,
    url:        `https://x.com/${d.user?.screen_name}/status/${d.id_str}`,
  }
}

export async function tweetDownload(url) {
  const info = await tweetInfo(url)
  const videos = info.medias.filter(m => m.type === 'video' || m.type === 'gif')
  const photos = info.medias.filter(m => m.type === 'photo')

  if (!videos.length && !photos.length)
    throw new Error('Este tweet no tiene media descargable')

  return {
    id:        info.id,
    text:      info.text,
    author:    info.author,
    createdAt: info.createdAt,
    videos,
    photos,
    hasVideo:  videos.length > 0,
    hasPhoto:  photos.length > 0,
  }
}
