// @soyaxell09/zenbot-scraper/src/scrapers/youtube.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios'

const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'
const delay = ms => new Promise(r => setTimeout(r, ms))

async function safeGet(url, config = {}) {
  try {
    return await axios.get(url, config)
  } catch (e) {
    const proxies = [
      `https://api.codetabs.com/v1/proxy?quest=${encodeURIComponent(url)}`,
      `https://api.allorigins.win/raw?url=${encodeURIComponent(url)}`
    ]
    for (const p of proxies) {
      try { return await axios.get(p, config) } catch (err) {}
    }
    throw e
  }
}

async function safePost(url, data, config = {}) {
  try {
    return await axios.post(url, data, config)
  } catch (e) {
    try {
      const p = `https://corsproxy.io/?${encodeURIComponent(url)}`
      return await axios.post(p, data, config)
    } catch (err) {
      throw e
    }
  }
}

function parseFileSize(size) {
  if (!size) return 0
  const units = { B: 1, KB: 1024, MB: 1024 ** 2, GB: 1024 ** 3, TB: 1024 ** 4 }
  const match = size.toString().trim().match(/([\d.]+)\s*(B|KB|MB|GB|TB)/i)
  if (!match) return 0
  return Math.round(parseFloat(match[1]) * (units[match[2].toUpperCase()] || 1))
}

function formatFileSize(bytes) {
  if (!bytes || isNaN(bytes)) return '0 B'
  const units = ['B', 'KB', 'MB', 'GB', 'TB']
  let i = 0
  while (bytes >= 1024 && i < units.length - 1) { bytes /= 1024; i++ }
  return `${bytes.toFixed(1).replace(/\.0$/, '')} ${units[i]}`
}

export async function getFileSize(url) {
  try {
    const res   = await axios.head(url, { timeout: 10000 })
    const bytes = parseInt(res.headers['content-length'] || 0)
    return formatFileSize(bytes)
  } catch { return '0 B' }
}

function normalizeYT(url) {
  try {
    const u = new URL(url)
    if (u.hostname.includes('youtu.be')) return url
    if (u.hostname.includes('youtube.com')) {
      if (u.pathname.includes('/watch'))   return `https://youtu.be/${u.searchParams.get('v')}`
      if (u.pathname.includes('/shorts/')) return `https://youtu.be/${u.pathname.split('/shorts/')[1]}`
      if (u.pathname.includes('/embed/'))  return `https://youtu.be/${u.pathname.split('/embed/')[1]}`
    }
    return url
  } catch { return url }
}

async function waitForDownload(mediaUrl) {
  if (!mediaUrl) throw new Error('No se recibió la URL de conversión')
  
  for (let i = 0; i < 20; i++) {
    try {
      const { data } = await safeGet(mediaUrl, { headers: { 'User-Agent': UA }, timeout: 15000 })
      if (data?.percent === 'Completed' && data?.fileUrl && data.fileUrl !== 'In Processing...') {
        return data.fileUrl
      }
    } catch {}
    await delay(3000)
  }
  throw new Error('El servidor YTDown tardó demasiado en generar la descarga')
}

async function fetchYtdownto(url) {
  const { data } = await safePost(
    'https://app.ytdown.to/proxy.php',
    new URLSearchParams({ url }).toString(),
    {
      headers: { 'Content-Type': 'application/x-www-form-urlencoded', 'User-Agent': UA },
      timeout: 20000,
    }
  )

  const api = data?.api || data
  if (!api || api.status === 'ERROR') throw new Error(api?.message || 'No se pudo obtener información del video de YTDown')

  let rawItems = data?.mediaItems || api?.medias || api?.links || []
  
  if (!Array.isArray(rawItems)) {
      rawItems = Object.values(rawItems)
  }

  if (rawItems.length === 0) {
      throw new Error('YTDown no devolvió calidades disponibles para este video (posible restricción regional).')
  }

  const qualities = rawItems.map((v, i) => {
    const type  = (v.type || v.mediaType || 'video').toLowerCase()
    const ext   = (v.mediaExtension || v.extension || 'mp4').toUpperCase()
    const qual  = v.mediaQuality || v.quality || (type === 'audio' ? '128k' : '360p')
    const match = qual.match(/(\d+)p|(\d+)k/)
    const res   = match ? match[0] : qual

    return {
      id:       i + 1,
      type:     type,
      quality:  res,
      label:    `${ext} - ${qual}`,
      size:     v.mediaFileSize || v.size || '0 B',
      sizeB:    parseFileSize(v.mediaFileSize || v.size),
      mediaUrl: v.mediaUrl || v.url || v.convertUrl,
      duration: v.mediaDuration || v.duration || '00:00',
    }
  })

  return { api, qualities }
}

export async function ytDownload(url, type = 'video', quality = '360p') {
  url = normalizeYT(url)

  const { api, qualities } = await fetchYtdownto(url)

  const isAudio  = type === 'mp3' || type === 'audio'
  const targetQ  = quality.toLowerCase()

  const filtered = isAudio
    ? qualities.filter(v => v.type === 'audio' || v.quality?.includes('k'))
    : qualities.filter(v => v.type === 'video' || v.quality?.includes('p'))

  if (!filtered.length) {
    const disponibles = qualities.map(v => v.quality).filter(Boolean).join(', ')
    throw new Error(`Calidad ${quality} no disponible. Disponibles: ${disponibles}`)
  }

  let selected
  if (isAudio) {
    const mp3s = filtered.filter(v => v.label?.toLowerCase().includes('mp3'))
    const pool = mp3s.length ? mp3s : filtered
    const exact = pool.find(v => v.quality?.toLowerCase() === targetQ)
    if (exact) {
      selected = exact
    } else {
      selected = pool.sort((a, b) => {
        const qa = parseInt(a.quality) || 0
        const qb = parseInt(b.quality) || 0
        return qb - qa
      })[0]
    }
  } else {
    selected = filtered.find(v => v.quality?.toLowerCase() === targetQ) || filtered[0]
  }

  if (!selected || !selected.mediaUrl) {
    throw new Error(`No se pudo extraer el enlace de conversión para la calidad solicitada.`)
  }

  const dlUrl = await waitForDownload(selected.mediaUrl)

  return {
    title:    api.title || 'YouTube Media',
    uploader: api.userInfo?.name || api.author || '',
    views:    api.mediaStats?.viewsCount || api.views || '',
    thumb:    api.imagePreviewUrl || api.thumbnail || '',
    type:     isAudio ? 'audio' : 'video',
    quality:  selected.quality,
    size:     selected.size,
    sizeB:    selected.sizeB,
    duration: selected.duration,
    url:      dlUrl,
  }
}

export async function ytInfo(url) {
  url = normalizeYT(url)
  const { api, qualities } = await fetchYtdownto(url)
  const duration = qualities[0]?.duration || ''
  const id = url.match(/youtu\.be\/([a-zA-Z0-9_-]{11})/)?.[1] || url.match(/v=([a-zA-Z0-9_-]{11})/)?.[1] || ''
  
  return {
    id,
    title:    api.title || 'YouTube Video',
    uploader: api.userInfo?.name || api.author || '',
    views:    api.mediaStats?.viewsCount || api.views || '',
    thumb:    api.imagePreviewUrl || api.thumbnail || '',
    duration,
    qualities,
  }
}

const INNERTUBE_URL = 'https://www.youtube.com/youtubei/v1'
let _config = null

async function getConfig() {
  if (_config) return _config
  const res = await axios.get('https://www.youtube.com/', {
    headers: {
      'User-Agent': UA,
      'Accept-Language': 'en-US,en;q=0.9',
      'Accept': 'text/html,application/xhtml+xml',
    },
    timeout: 15000,
  })
  const html          = res.data
  const key           = html.match(/"INNERTUBE_API_KEY"\s*:\s*"([^"]+)"/)?.[1] || 'AIzaSyA8eiZmM1FaDVjRy-df2KTyQ_vz_yYM394'
  const visitorData   = html.match(/"visitorData"\s*:\s*"([^"]+)"/)?.[1] || ''
  const clientVersion = html.match(/"clientVersion"\s*:\s*"([^"]+)"/)?.[1] || '2.20240101.00.00'
  _config = { key, visitorData, clientVersion }
  return _config
}

export async function ytSearch(query, limit = 5) {
  const cfg = await getConfig()
  const res = await axios.post(
    `${INNERTUBE_URL}/search?key=${cfg.key}&prettyPrint=false`,
    {
      query,
      context: {
        client: {
          clientName: 'WEB',
          clientVersion: cfg.clientVersion,
          hl: 'en', gl: 'US',
          visitorData: cfg.visitorData,
        },
      },
    },
    {
      headers: {
        'User-Agent': UA,
        'X-YouTube-Client-Name': '1',
        'X-YouTube-Client-Version': cfg.clientVersion,
        'Content-Type': 'application/json',
        'X-Goog-Visitor-Id': cfg.visitorData,
      },
      timeout: 15000,
    }
  )

  const contents =
    res.data?.contents
      ?.twoColumnSearchResultsRenderer
      ?.primaryContents
      ?.sectionListRenderer
      ?.contents?.[0]
      ?.itemSectionRenderer
      ?.contents || []

  const results = []
  for (const item of contents) {
    if (results.length >= limit) break
    const v = item?.videoRenderer
    if (!v?.videoId) continue
    results.push({
      id:        v.videoId,
      title:     v.title?.runs?.[0]?.text || '',
      url:       `https://www.youtube.com/watch?v=${v.videoId}`,
      thumbnail: v.thumbnail?.thumbnails?.at(-1)?.url || '',
      duration:  v.lengthText?.simpleText || '',
      views:     v.viewCountText?.simpleText || '',
      channel:   v.ownerText?.runs?.[0]?.text || '',
      published: v.publishedTimeText?.simpleText || '',
    })
  }

  if (!results.length) throw new Error('Sin resultados de búsqueda')
  return results
}
