// @soyaxell09/zenbot-scraper/src/nsfw/rule34.js
/*
 * © Created by AxelDev09 🔥
 * GitHub: https://github.com/AxelDev09
 * Instagram: @axeldev09
 * Deja los créditos we 🗣️
 */

import axios from 'axios';
import * as cheerio from 'cheerio';

const UA = 'Mozilla/5.0 (Linux; Android 11; Redmi Note 8) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36';

function parsePaheal(xml) {
  const $ = cheerio.load(xml, { xmlMode: true });
  const items = [];
  $('tag').each((_, el) => {
    const file_url = $(el).attr('file_url') || '';
    if (!file_url) return;
    items.push({
      id:      $(el).attr('id'),
      url:     file_url,
      full:    file_url,
      preview: $(el).attr('preview_url') || '',
      tags:    $(el).attr('tags') || '',
      score:   parseInt($(el).attr('score')) || 0,
      width:   parseInt($(el).attr('width')) || 0,
      height:  parseInt($(el).attr('height')) || 0
    });
  });
  return items;
}

export async function rule34Random(tags = null) {
  for (let intento = 0; intento < 3; intento++) {
    const pid    = Math.floor(Math.random() * 5);
    const params = { limit: 20, pid };
    if (tags) params.tags = tags;
    const { data } = await axios.get('https://rule34.paheal.net/api/danbooru/find_posts', {
      params,
      headers: { 'User-Agent': UA },
      timeout: 12000
    });
    const items = parsePaheal(data);
    if (items.length) return items[Math.floor(Math.random() * items.length)];
  }
  throw new Error('No se encontraron resultados.');
}
